### PR TITLE
added shopping cart and cart items functional

### PR DESCRIPTION
**Pull request by luk14n**

### CODE REVIEW

File: src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+     * After that item is saved and resides in needed cart.
+     *
+     * @param userId         ID of user whose cart it is
+     * @param requestDto DTO of requested item to add to the cart
+     * @return DTO of updated cart (after desired item is added)
+     * @throws RuntimeException if the user's cart cannot be found in the DB.
+     */
+    @Override
+    public ShoppingCartDto addBooksToCartByUserId(Long userId, AddItemToCartRequestDto requestDto) {
+        ShoppingCart cartFromDB = getCartFromDB(userId);
```
Comment (koval-kn) [inline src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java:39]: if cart is not found we should create a new one, not throw exception

File: src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        CartItem requestedItemToAdd = itemMapper.toModel(requestDto);
```
Comment (koval-kn) [inline src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java:40]: book is not added, get it by id from DB and set properly

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -93,8 +106,15 @@ private CartItem getItemFromDb(Long cartItemId) {
...
+        Optional<ShoppingCart> cartFromDb = cartRepository.findByUserId(userId);
+        if (cartFromDb.isEmpty()) {
+            ShoppingCart shoppingCart = new ShoppingCart();
+            User userFromDb = userRepository.findById(userId)
+                    .orElseThrow(() -> new EntityNotFoundException(
+                            "User not found by user id: " + userId));
+            shoppingCart.setUser(userFromDb);
+            return cartRepository.save(shoppingCart);
+        }
+        return cartFromDb.get();
```
Comment (Vasul-Pankiv) [inline src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java]: you can simplify it. Find the users first and then use orElseGet() to get or create shopping cart

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/com/lukian/bookstore/model/ShoppingCart.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.OneToMany;
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import java.util.Set;
+import lombok.Data;
+
+@Entity
+@Data
+@Table(name = "shopping_carts")
+public class ShoppingCart {
```
Comment (Rommelua) [inline src/main/java/com/lukian/bookstore/model/ShoppingCart.java:21]: Don't forget to implement soft delete approach

File: src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository cartRepository;
+    private final CartItemRepository itemRepository;
+    @Qualifier("cartItemMapper")
+    private final CartItemMapper itemMapper;
+    @Qualifier("shoppingCartMapper")
+    private final ShoppingCartMapper cartMapper;
```
Comment (Rommelua) [inline src/main/java/com/lukian/bookstore/service/cart/ShoppingCartServiceImpl.java]:
```suggestion
    private final CartItemMapper itemMapper;
    private final ShoppingCartMapper cartMapper;
```

File: src/main/java/com/lukian/bookstore/service/user/UserServiceImpl.java [hunk 6]
```diff
@@ -13,14 +20,39 @@
...
     }
+
+    private void assignDefaultUserRoleToNewUser(User user) {
+        Role roleFromDb = roleRepository.findRoleByRole(Role.RoleName.ROLE_USER)
+                        .orElseThrow(() -> new EntityNotFoundException(
+                                "Cannot find role by given name"));
+        user.getRoles().add(roleFromDb);
+        HashSet<Role> roles = new HashSet<>();
+        roles.add(roleFromDb);
+        user.setRoles(roles);
```
Comment (Rommelua) [inline src/main/java/com/lukian/bookstore/service/user/UserServiceImpl.java]:
```suggestion
```

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**